### PR TITLE
[ML] Refine client code for data frame analytics

### PR DESCRIPTION
**Pull request by przemekwitek**
Make the client code more usable and uniform:
 * Introduce Builder classes
 * Use Strings.toString consistently
 * Remove unused copy constructors

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/OutlierDetection.java [hunk 1]
```diff
@@ -36,40 +35,42 @@
 public class OutlierDetection implements DataFrameAnalysis {
 
     public static OutlierDetection fromXContent(XContentParser parser) {
-        return PARSER.apply(parser, null);
+        return PARSER.apply(parser, null).build();
+    }
+
+    public static OutlierDetection getDefaultInstance() {
```
Comment (dimitris-athanasiou) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/OutlierDetection.java]: rename to just `default`? `getDefaultInstance` could imply the same instance is always returned, which is not the case.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/MlDataFrameAnalysisNamedXContentProvider.java [hunk 2]
```diff
@@ -21,19 +21,17 @@
 import org.elasticsearch.common.xcontent.NamedXContentRegistry;
 import org.elasticsearch.plugins.spi.NamedXContentProvider;
 
-import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 
 public class MlDataFrameAnalysisNamedXContentProvider implements NamedXContentProvider {
```
Comment (dimitris-athanasiou) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/MlDataFrameAnalysisNamedXContentProvider.java]: Do we need this class at all in the client side?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/OutlierDetection.java [hunk 3]
```diff
@@ -78,11 +79,21 @@ public OutlierDetection(@Nullable Integer nNeighbors, @Nullable Method method) {
...
-     */
-    public OutlierDetection() {
-        this(null, null);
+    @Override
+    public String getName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    public Map<String, Object> getParams() {
```
Comment (dimitris-athanasiou) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/OutlierDetection.java]: This method is only necessary for the backend. We can remove it from `DataFrameAnalysis`.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/DataFrameAnalyticsConfig.java [hunk 4]
```diff
@@ -178,23 +169,6 @@ public String toString() {
         private FetchSourceContext analyzedFields;
         private ByteSizeValue modelMemoryLimit;
 
-        private Builder() {}
```
Comment (dimitris-athanasiou) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/DataFrameAnalyticsConfig.java]: As we have `builder()` methods to the respective classes, I wonder if we should have private constructors for the builders. Doing so would mean users would have a single way to access the builders. I am not quite sure if this makes sense, but it feels like it'd force users to write more consistent code.

**Review by dimitris-athanasiou — COMMENTED** [review]
Looks good. Left a few comments.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/OutlierDetection.java (hunk 3 above)
Comment (przemekwitek) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/OutlierDetection.java]: Done.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/DataFrameAnalyticsConfig.java (hunk 4 above)
Comment (przemekwitek) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/DataFrameAnalyticsConfig.java]: Definitely a good idea! Done.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/MlDataFrameAnalysisNamedXContentProvider.java (hunk 2 above)
Comment (przemekwitek) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/MlDataFrameAnalysisNamedXContentProvider.java]: I believe we do need it at least for AbstractXContentTestCase auto-generated tests. Without this class, the parser won't know how to parse the JSON when it encounters "outlier_detection".

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/OutlierDetection.java (hunk 1 above)
Comment (przemekwitek) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/OutlierDetection.java]: Renamed to "createDefault" as "default" is a keyword in Java.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Pull request merged**